Complete default blocks

jcdyer opened the pull request:
Add code to submit completions from verticals on behalf of descendent blocks that aren't scorable and don't implement a custom completion method.

**JIRA tickets**: N/A (Part of [Completion API](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/162247762/Completion+API) work)

**Discussions**: See Completion API document above.

**Dependencies**: #16112 #16047 

**Screenshots**: N/A

**Sandbox URL**: Upon request.

**Merge deadline**: None

**Testing instructions**:

1. Create a waffle switch named `completion.enable_completion_tracking` with is_active set to True.
2. Find a user that has no completions in the demo course.
       from django.contrib.auth.models import User
       from lms.djangoapps.completion.models import BlockCompletion
       pick_from = [u for u in User.objects.all() if not BlockCompletion.objects.filter(user=u).exists()]
3. Register for the demo course (if needed)
4. Visit "Example Week 1: Getting Started -> Lesson 1 - Getting Started -> Getting Started"
5. Once the page has fully rendered, verify that `BlockCompletion.objects.filter(user=user, course_key=CourseKey.from_string('course-v1:edX+DemoX+Demo_course')` returns a completion object with a completion value of 1.0 for the HTML block and Discussion block on that page. (A later ticket will exclude discussion blocks from completion tracking).
6. Visit the second-to-last vertical in the sequential: 3. Visit "Example Week 1: Getting Started -> Lesson 1 - Getting Started -> Interactive questions".
7. Completion of scorable XBlocks is handled separately.  Verify that visiting the page doesn't create a  completion for the capa problem on the page.

**Author notes and concerns**:

1. There may be more work that could be done to extract this into a CompletionContainerMixin, so implementations could add it to a different block type than verticals if appropriate for their installation, but that feels like an optimization that should be deferred.
2. I considered pushing the functionality to submit completions to the runtime service as well, but since the runtime itself is needed to publish events, and the runtime is directly available to the block, it didn't seem necessary.  
3. Is there a better way to include the runtime service than hard coding it on the LMS runtime, and making it optional on the vertical?  
4. This will need a follow-up to be more specific about when to complete blocks.  It currently marks them complete when the vertical is first visited, but it should do it when the block has has been viewed, which will require checking coordinates of the block's containing element in the viewport.  We initially said that it should be when the block 
5. I tried to convert the vertical file to ES6, as suggested by the front-end guidelines, but I couldn't make the import of the bookmark button work.  I tried several ways of specifying the path, and asked on the openedx slack instance, but couldn't get it working.  I don't think it's worth me putting any more time in, but I imagine it would be pretty straightforward for someone with more javascript understanding.  

**Reviewers**
- [x] @mtyaka / @bradenmacdonald 
- [x] edX reviewer @robrap 

FYI: @e-kolpakov 

**Settings**: N/A

**TODO**:
- [x] Squash commits
- [x] Create a follow-up ticket for improving completion timing (See Author notes \#4 above).

edx-webhook commented:
Thanks for the pull request, @jcdyer! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=16234&repo=edx%2Fedx-platform"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

jcdyer commented:
@bradenmacdonald I thought so, but I was getting lint errors about `const` and `let` being reserved keywords.  

CC @andy-armstrong What's the state of ES6 in edx-platform?

arizzitano commented:
@jcdyer @bradenmacdonald is right -- es6/es2015 should be used for new js development. You can't mix es5 and es2015 within the same file, though, so if you are unable to convert the entire file to an es2015 module (transpiled through webpack) it may be a moot point.

jcdyer commented:
Thanks @arizzitano . Converting the whole file shouldn't be hard, though I've got a follow-up question about how to configure it on slack: https://openedx.slack.com/archives/C0EUBSV7D/p1508347878000087.

jcdyer commented:
jenkins run python

robrap commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -1,16 +1,47 @@
+  }
+
+  static registerCompletions(runtime, element) {
+    const $verts = $(element).find('.vert');
```
Maybe rename `$verts` to `$blocks_in_vertical`?

robrap commented on lms/templates/vert_module.html:
```diff
@@ -10,7 +10,11 @@ <h2 class="hd hd-2 unit-title">${unit_title}</h2>
-  <div class="vert vert-${idx}" data-id="${item['id']}">
+<div class="vert vert-${idx}" data-id="${item['id']}" \
+% if item['id'] in watched_completables:
+  data-completable \
```
`data-completable-by-viewing` is long, but I find it less confusing than if one were trying to understand why other completable blocks don't seem to be completable.

robrap commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -1,16 +1,47 @@
+  static registerCompletions(runtime, element) {
+    const $verts = $(element).find('.vert');
+    $verts.each(
+      (idx, completable) => {
```
Maybe renaming `completable` to `block` would make this less confusing?

robrap commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -1,16 +1,47 @@
+    const $verts = $(element).find('.vert');
+    $verts.each(
+      (idx, completable) => {
+        if (completable.dataset.completable === undefined) {
```
Two possible (untested) name changes captured below. 
`completable.dataset.completable` => `block.dataset.completableByViewing`

See [stackoverflow post](https://stackoverflow.com/questions/22753629/jquery-get-html-5-data-attributes-with-hyphens-and-case-sensitivity) for where I took naming conversion from `data-completable-by-viewing` => `dataset.completableByViewing`.

robrap commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -40,6 +40,7 @@
 
 class SequenceFields(object):
     has_children = True
+    completion_method = 'aggregator'
```
Can we have a string constant somewhere?  Also, not seeing this used anywhere.

jcdyer replied:
There will be a string constant in the XBlock repo.  I'll update this to use that once it lands (possibly a later commit, depending on how long it takes to get this one merged).

robrap commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -1,28 +1,52 @@
+    This is true of any non-customized, non-scorable, completable block.
+    """
+    return (
+        getattr(block, 'completion_method', 'completable') == 'completable'
```
Also, I thought I saw @bradenmacdonald comment somewhere on `'completable'` being a named constant, but not seeing that, so just noting it again.

jcdyer replied:
This is a const-soon.  See above.

jcdyer replied:
@robrap Here's the -soon joke.

robrap replied:
Ah yes.  Thank you again.  I like.

robrap commented on common/lib/xmodule/xmodule/vertical_block.py:
```diff
@@ -37,6 +61,21 @@ class VerticalBlock(SequenceFields, XModuleFields, StudioEditableBlock, XmlParse
+        if completion_service is None:
+            return set()
+        # pylint: disable=no-member
+        completables = {block.location for block in self.get_display_items() if is_completable_by_viewing(block)}
```
Same questions/concerns as earlier about rollout.  Can this be waffle protected?  Maybe we can discuss elsewhere.

robrap commented on lms/djangoapps/completion/services.py:
```diff
@@ -0,0 +1,49 @@
+"""
```
Note to self: This is where I am leaving off for now because: 1) I have limited time allocated to this and 2) I may have jumped into this review prematurely.  I'm getting confused around what reviews are needed when, and I should probably wait for the explicit messages from @jcdyer asking for review. :)

jcdyer replied:
I'm having trouble juggling all the PRs right now, too.  I'm working toward getting the early ones merged, so there's less to focus on.

robrap commented:
@jcdyer: As a reminder, I did not review all the code.  Please leave an explicit comment when and if @bradenmacdonald and I should review like you did on the others.  Thanks.

jcdyer commented:
This is ready for an (opencraft) review.  @mtyaka you are listed as reviewer on this, but you signed up for that in a different sprint, so I understand if you don't have time for reviewing this.  I'll let you and @bradenmacdonald decide how to handle this.

@robrap You can hold off until I get thumbs from one of the above, if you prefer.

jcdyer commented:
@mtyaka @bradenmacdonald Cancel that earlier request.  It looks like I'm missing some test coverage on the new waffle flag.  Let me get that in place, and I'll update with a new review request.

jcdyer commented:
@bradenmacdonald @mtyaka  This is ready for review by one of you.

jcdyer commented:
There were some bokchoy errors, because I didn't realize that library content is rendered using the vert_module.html template.  I set it up not to track completion.  We can discuss if that needs to be handled differently, and perhaps create a new ticket if it does.  I suspect the library content module will need to be treated as a custom completable, and track completion of its children independently.

bradenmacdonald commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -13,4 +13,23 @@ window.VerticalStudentView = function(runtime, element) {
+                        block_key: blockKey,
+                        completion: 1.0
+                    })
+                });
```
If I view a sequential like "Lesson 1 - Getting Started" in the demo course and click from the first vertical to another vertical and then back to the first, I see that it still POSTs a `publish_completion` event to the server - it has "forgotten" that it already marked the block as complete.

I was going to suggest calling `delete block.dataset.completableByViewing;` after the AJAX request completes successfully, but I think that each vertical gets rendered from an HTML string when switching verticals in a sequential, so that approach wouldn't work.

If you think it's worth doing, maybe add a global variable to track already-submitted completion events? Or we can ignore the duplicates, as it's not a huge optimization.

robrap commented:
I have not yet reviewed this, but just wanted to leave some high-level comments:
1. I imagine that edX is going to need a batch call before this can be turned on because there are some courses with huge verticals.
2. There was discussion of a timer before marking (not sure if that was implemented or future).  For future required features, it might be helpful to have a TODO with a description of what is coming and/or a link to a Confluence page or JIRA ticket.
3. For scrolling/focus based events, would each one have a timer or would a quick scroll mark things as complete?  This will relate both to the need for batch eventing and/or the implementation of batch eventing.  Also, I'm not sure if trying to time this details with also having a UI will work, unless you want to add another flag for whether or not to do this.

One of the assumptions I had was that the Completion API would be implemented and merged such that it could be turned on for edX.org without performance issues.  It is ok if work is being chunked to get smaller PRs.  But I also want to ensure that some details won't get dropped.

Thanks.

jcdyer commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -514,6 +514,7 @@ def handle_completion_event(block, event):
         if not completion_waffle.waffle().is_enabled(completion_waffle.ENABLE_COMPLETION_TRACKING):
             raise Http404
         else:
+            log.info("Completing block %s", block.location)
```
Yes.  I've removed them, and included a comment to explain the early return.

jcdyer commented:
@bradenmacdonald Some of your comments were attached to a particular commit.  I rebased against the latest master to resolve merge issues & those comments (about the vertical_student_view.js file) got lost, but I've addressed the issues you mentioned.  Ready for another round.

jcdyer commented:
jenkins run bokchoy

jcdyer commented:
Failure due to: https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/45642/

Trying rerun.

bradenmacdonald commented:
@jcdyer I just gave this another spin. It seems to be working very well; there's no duplicate events when changing tabs in the sequential, nor are there repeat events when refreshing the whole page.

Code in your recent commits looks good to me too. Thanks!

So my :+1: stands.

jcdyer commented:
Great!  Thanks, @bradenmacdonald .  

@robrap This is now ready for your review.

robrap commented:
@jcdyer: I'm working with the team to get this work assigned and scheduled.  I will let you know as soon as I hear.

robrap commented on lms/templates/vert_module.html:
```diff
@@ -10,7 +10,11 @@ <h2 class="hd hd-2 unit-title">${unit_title}</h2>
 <div class="vert-mod">
 % for idx, item in enumerate(items):
-  <div class="vert vert-${idx}" data-id="${item['id']}">
+<div class="vert vert-${idx}" data-id="${item['id']}" \
```
Can you add back indent?

robrap commented on requirements/edx/base.txt:
```diff
@@ -203,4 +203,4 @@ py2neo==3.1.2
 # Support for plugins
 web-fragments==0.2.2
-xblock==1.0.0
+git+https://github.com/edx/XBlock@xblock-1.1.0#egg=XBlock==1.1.0
```
Working on updating.  Will be 1.1.1.

robrap replied:
https://pypi.python.org/pypi/XBlock/1.1.1 is available.

jcdyer commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -1,4 +1,14 @@
+// Using a global variable for SEEN_COMPLETABLES because this needs to track
+// completables between separate loads of the same vertical (when a learner
+// goes from one tab to the next, and then navigates back within a given
+// sequential).
```
Copy pasting this with a slight tweak to the first sentence to simplify the grammatical structure.

jcdyer commented:
@robrap.  Ready for another look.

jcdyer commented:
Tests don't seem to be running right now.  I'll try to get them running again in an hour or so. (FYI @edx/testeng)

jcdyer commented:
Tests don't seem to be running right now.  I'll try to get them running again in an hour or so. (FYI @edx/testeng)

michaelyoungstrom commented:
Sorry @jcdyer , we've been having issues with Jenkins over the last 2 days. Your tests are running right now, and we'll continue monitoring.

robrap commented on common/lib/xmodule/xmodule/assets/vertical/public/js/vertical_student_view.js:
```diff
@@ -13,4 +24,32 @@ window.VerticalStudentView = function(runtime, element) {
+            if (block.dataset.completableByViewing === undefined) {
+                return;
+            }
+            // TODO:
```
@jcdyer: Thanks.   Can you please change this first line to when you squash?:
```
// TODO: EDUCATOR-1778
```
In the future, I may ask you to use "TODO: EDUCATOR-1778" so we can track blockers to enabling for edx.org.  Thank you!

jcdyer replied:
Great idea.

pomegranited commented:
@jcdyer Is there documentation for the edx-platform waffle flags anywhere?  Would be worth updating it to include this new one.

jcdyer commented:
@pomegranited I actually don't know if there is a central documentation source for the waffle flags and switches, though I agree it's a little hard to work with them as the prefixing makes it so the full switch name doesn't actually show up anywhere. I'll at least as the full name in a comment on the switch itself, and if @robrap knows of a place to document it, I'll add it there too.
